File: examples/seller-agent/src/seller_agent/settlement.py
"""EIP-712 settlement profile — a self-contained implementation of spec §4.4.

The coordination layer carries `agreementSig` and the payload `*Sig` members
as opaque strings, but an independent agent must PRODUCE them, and §4.4 makes
their construction normative: two domains, eight frozen struct type strings,
and a derivation rule for every field. This module implements exactly that —
from the published document, with the same generic crypto libraries the rest
of the example uses (coincurve for recoverable secp256k1, pycryptodome for
keccak-256), and no Kite code. `tests/test_settlement_vectors.py` replays the
published `vectors/v1/settlement` set against it, which is what makes this
file an implementation of the spec rather than of the Kite engine.

Chain parameters are deployment configuration. The defaults are the PUBLISHED
FIXTURE values the settlement vectors are minted with, so the demo's
signatures are verifiable against the vectors; a real deployment sets both.
"""

from __future__ import annotations

import os
from decimal import Decimal, InvalidOperation

from coincurve import PrivateKey, PublicKey
from Crypto.Hash import keccak

CHAIN_ID = int(os.environ.get("KITE_CHAIN_ID", "2368"))
VAULT_ADDRESS = os.environ.get("KITE_VAULT_ADDRESS", "0x" + "ec" * 20)

# §4.4 struct type strings — verbatim, frozen.
AGREEMENT_TYPE = (
    "Agreement(bytes32 agreementId,bytes32 termsHash,uint256 amount,"
    "address buyerAgent,address sellerAgent)"
)
ACTIVATION_TYPE = (
    "Activation(bytes32 termsHash,address buyer,address buyerAgent,"
    "address sellerAgent,address sellerPayout,address arbiter,uint256 amount,"
    "uint64 fundingDeadline,uint64 deliveryWindow,uint64 deliveryConfirmationWindow,"
    "uint64 appealResponseWindow,uint64 arbitrationWindow)"
)
DELIVERY_TYPE = (
    "Delivery(bytes32 dealId,bytes32 termsHash,bytes32 deliveryHash,"
    "bytes32 receiptHash,uint64 nonce,uint64 expiry)"
)
ACCEPTANCE_TYPE = (
    "Acceptance(bytes32 dealId,bytes32 termsHash,bytes32 receiptHash,"
    "uint64 nonce,uint64 expiry)"
)
REJECTION_TYPE = (
    "Rejection(bytes32 dealId,bytes32 termsHash,bytes32 reasonHash,"
    "bytes32 receiptHash,uint64 nonce,uint64 expiry)"
)
APPEAL_TYPE = (
    "Appeal(bytes32 dealId,bytes32 termsHash,bytes32 receiptHash,"
    "uint64 nonce,uint64 expiry)"
)
REFUND_CONSENT_TYPE = (
    "RefundConsent(bytes32 dealId,bytes32 termsHash,bytes32 receiptHash,"
    "uint64 nonce,uint64 expiry)"
)
RESOLUTION_TYPE = (
    "Resolution(bytes32 dealId,bytes32 termsHash,bytes32 decisionId,"
    "uint16 sellerBps,bytes32 receiptHash,uint64 nonce,uint64 expiry)"
)

# A demo stand-in for §4.4's receiptHash anchor (the PRIOR transition-proof
# hash, read via `status`). With no Runtime there is no proof chain; the zero
# anchor keeps the construction real and the value visibly a placeholder.
ZERO_ANCHOR = "0x" + "00" * 32


def _keccak(data: bytes) -> bytes:
    h = keccak.new(digest_bits=256)
    h.update(data)
    return h.digest()


# ── ABI words (every §4.4 field encodes as one 32-byte word) ─────────────────

def word_int(n: int) -> bytes:
    return int(n).to_bytes(32, "big")


def word_bytes32(ref: str) -> bytes:
    """A 32-byte value in either spelling the system mints: sha256:<hex> or
    0x<hex>. Exactly 32 bytes — nothing padded or truncated."""
    h = ref.removeprefix("sha256:").removeprefix("0x")
    if len(h) != 64:
        raise ValueError(f"not a 32-byte value: {ref}")
    return bytes.fromhex(h)


def word_addr(addr: str) -> bytes:
    return b"\x00" * 12 + bytes.fromhex(addr.removeprefix("0x"))


def bytes32_equal(a: str, b: str) -> bool:
    """Whether two 32-byte references name the SAME bytes, whatever their
    spelling. The system legitimately mints both forms — the coordination
    layer says `sha256:<hex>`, the chain-adjacent views say `0x<hex>` — and
    word_bytes32 already treats them as one value for SIGNING. Comparing the
    strings instead of the bytes rejects a value against itself: the engine's
    funding view spells the termsHash `0x…` while the accepted contract's
    anchor is `sha256:…`, and a literal comparison refuses every deal.
    Unparseable input is False, not an exception — callers are validating."""
    try:
        return word_bytes32(a) == word_bytes32(b)
    except (ValueError, TypeError):
        return False


# ── §4.4 field derivations ───────────────────────────────────────────────────

def usdc_base_units(amount: str) -> int:
    """`price.amount` ("25.00") as the uint256 the Agreement digest commits:
    USDC base units, 6 decimals. An unparseable or over-precise amount is an
    ERROR — a silent zero would verify a signature over the wrong amount.

    For the CONTRACT's decimal spelling only. The wire Activation's `amount`
    is already base units — see base_units below; running it through this
    conversion again multiplies it by 10^6 and every Activation signature
    breaks."""
    try:
        units = Decimal(amount) * (10**6)
    except InvalidOperation as exc:
        raise ValueError(f"not a decimal amount: {amount!r}") from exc
    if units != int(units) or units <= 0:
        raise ValueError(f"not a positive USDC amount at 6 decimals: {amount!r}")
    return int(units)


def base_units(amount: str) -> int:
    """The wire Activation's `amount`: ALREADY base units, a bare integer
    decimal string ("25000000" — activation.schema.json pins the ^[0-9]+$
    pattern, and the engine parses it with big.Int.SetString, no conversion).
    Distinct from usdc_base_units on purpose: the two `uint256 amount` fields
    have different SOURCES — the Agreement's comes from the contract's decimal
    price, the Activation's from the Runtime's already-converted view — and
    collapsing them is exactly the double-conversion this split prevents."""
    amount = amount.strip()
    if not amount.isdigit():
        raise ValueError(f"not a base-units integer string: {amount!r}")
    units = int(amount)
    if units <= 0:
        raise ValueError(f"not a positive base-units amount: {amount!r}")
    return units


def agreement_id32(deal_id: str) -> bytes:
    """bytes32 agreementId = keccak256(UTF-8 of the Runtime-assigned id)."""
    return _keccak(deal_id.encode())


def reason_hash(reason_code: str) -> str:
    """The vault's bytes32 reasonHash = keccak256(UTF-8 reasonCode) (§4.2)."""
    return "0x" + _keccak(reason_code.encode()).hex()


def decision_id32(decision_id: str) -> str:
    """The vault's bytes32 decisionId = keccak256(UTF-8 decisionId) (§4.2)."""
    return "0x" + _keccak(decision_id.encode()).hex()


def demo_deal_id32(deal_id: str) -> str:
    """DEMO ONLY. The real bytes32 dealId is EscrowVault.dealIdFor(Activation),
    READ BACK through the funding/status interactions (§4.4) — it cannot be
    derived without the Activation. With no Runtime to read from, the demo
    anchors its vault structs to keccak256 of the deal id string instead; the
    construction and verification around it are the real ones."""
    return "0x" + _keccak(deal_id.encode()).hex()


# ── domains ──────────────────────────────────────────────────────────────────

_DOMAIN_TYPE_WITH_CONTRACT = (
    "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
)
_DOMAIN_TYPE_NO_CONTRACT = "EIP712Domain(string name,string version,uint256 chainId)"


def vault_domain(chain_id: int = CHAIN_ID, vault: str = VAULT_ADDRESS) -> bytes:
    """KiteEscrowVault/1 — the domain of every payload *Sig (vault calls)."""
    return _keccak(
        _keccak(_DOMAIN_TYPE_WITH_CONTRACT.encode())
        + _keccak(b"KiteEscrowVault") + _keccak(b"1")
        + word_int(chain_id) + word_addr(vault)
    )


def agreement_domain(chain_id: int = CHAIN_ID) -> bytes:
    """KiteFulfill/1 — the Agreement (accept gate) domain. NO verifyingContract:
    the accept gate is off-chain, so there is no contract address to bind."""
    return _keccak(
        _keccak(_DOMAIN_TYPE_NO_CONTRACT.encode())
        + _keccak(b"KiteFulfill") + _keccak(b"1")
        + word_int(chain_id)
    )


def struct_hash(type_string: str, *words: bytes) -> bytes:
    return _keccak(_keccak(type_string.encode()) + b"".join(words))


def typed_digest(domain: bytes, shash: bytes) -> bytes:
    """EIP-712: keccak256( 0x1901 ‖ domainSeparator ‖ structHash )."""
    return _keccak(b"\x19\x01" + domain + shash)


# ── signing and recovery over a 32-byte digest ──────────────────────────────

def sign_digest32(priv: PrivateKey, digest: bytes) -> str:
    """Recoverable secp256k1 over an EIP-712 digest, 0x r‖s‖v hex, v ∈ {27,28}.
    Same key and encoding as the coordination profile (§8: one agent key)."""
    raw = priv.sign_recoverable(digest, hasher=None)
    return "0x" + raw[:64].hex() + bytes([raw[64] + 27]).hex()


def recover_digest32(digest: bytes, sig_hex: str) -> str:
    body = bytes.fromhex(sig_hex.removeprefix("0x"))
    if len(body) != 65 or body[64] not in (27, 28):
        raise ValueError("settlement signature must be 65 bytes with v in {27,28}")
    pub = PublicKey.from_signature_and_message(
        body[:64] + bytes([body[64] - 27]), digest, hasher=None
    )
    return "0x" + _keccak(pub.format(compressed=False)[1:])[-20:].hex()


def verifies(digest: bytes, sig_hex: str, address: str) -> bool:
    try:
        return recover_digest32(digest, sig_hex).lower() == address.lower()
    except (ValueError, Exception):  # noqa: B014 - any decode failure is a rejection
        return False


# ── §4.4 digests, one per struct ─────────────────────────────────────────────

def agreement_digest(
    deal_id: str, terms_hash_ref: str, amount: str,
    buyer_address: str, seller_address: str, chain_id: int = CHAIN_ID,
) -> bytes:
    """The digest behind `agreementSig` (§4.1/§4.4). Exists only once the
    Runtime has assigned deal_id — which is WHY agreementSig is two-phase."""
    return typed_digest(
        agreement_domain(chain_id),
        struct_hash(
            AGREEMENT_TYPE,
            agreement_id32(deal_id),
            word_bytes32(terms_hash_ref),
            word_int(usdc_base_units(amount)),
            word_addr(buyer_address),
            word_addr(seller_address),
        ),
    )


def activation_digest(
    terms_hash_ref: str, buyer_wallet: str, buyer_agent: str, seller_agent: str,
    seller_payout: str, arbiter: str, amount: str, windows: dict[str, int],
    chain_id: int = CHAIN_ID, vault: str = VAULT_ADDRESS,
) -> bytes:
    """The digest behind BOTH funding artifacts (§6.2.1): buyerActivationSig
    and sellerActivationSig sign this same value, and the vault's fund()
    verifies the pair. EVERY member — including `amount` — is the value read
    back through the `funding` interaction, and the read-back amount is
    ALREADY base units ("25000000", not "25.00"): converting it again would
    sign a digest 10^6 too large, refused only at the vault."""
    return typed_digest(
        vault_domain(chain_id, vault),
        struct_hash(
            ACTIVATION_TYPE,
            word_bytes32(terms_hash_ref), word_addr(buyer_wallet),
            word_addr(buyer_agent), word_addr(seller_agent),
            word_addr(seller_payout), word_addr(arbiter),
            word_int(base_units(amount)),
            word_int(windows["fundingDeadline"]), word_int(windows["deliveryWindow"]),
            word_int(windows["deliveryConfirmationWindow"]),
            word_int(windows["appealResponseWindow"]), word_int(windows["arbitrationWindow"]),
        ),
    )


def delivery_digest(
    deal_id32: str, terms_hash_ref: str, delivery_hash_ref: str,
    receipt_hash: str = ZERO_ANCHOR, nonce: int = 0, expiry: int = 0,
    chain_id: int = CHAIN_ID, vault: str = VAULT_ADDRESS,
) -> bytes:
    return typed_digest(
        vault_domain(chain_id, vault),
        struct_hash(
            DELIVERY_TYPE,
            word_bytes32(deal_id32), word_bytes32(terms_hash_ref),
            word_bytes32(delivery_hash_ref), word_bytes32(receipt_hash),
            word_int(nonce), word_int(expiry),
        ),
    )


def acceptance_digest(
    deal_id32: str, terms_hash_ref: str,
    receipt_hash: str = ZERO_ANCHOR, nonce: int = 0, expiry: int = 0,
    chain_id: int = CHAIN_ID, vault: str = VAULT_ADDRESS,
) -> bytes:
    return typed_digest(
        vault_domain(chain_id, vault),
        struct_hash(
            ACCEPTANCE_TYPE,
            word_bytes32(deal_id32), word_bytes32(terms_hash_ref),
            word_bytes32(receipt_hash), word_int(nonce), word_int(expiry),
        ),
    )


def rejection_digest(
    deal_id32: str, terms_hash_ref: str, reason_code: str,
    receipt_hash: str = ZERO_ANCHOR, nonce: int = 0, expiry: int = 0,
    chain_id: int = CHAIN_ID, vault: str = VAULT_ADDRESS,
) -> bytes:
    return typed_digest(
        vault_domain(chain_id, vault),
        struct_hash(
            REJECTION_TYPE,
            word_bytes32(deal_id32), word_bytes32(terms_hash_ref),
            word_bytes32(reason_hash(reason_code)), word_bytes32(receipt_hash),
            word_int(nonce), word_int(expiry),
        ),
    )


def resolution_digest(
    deal_id32: str, terms_hash_ref: str, decision_id: str, seller_bps: int,
    receipt_hash: str = ZERO_ANCHOR, nonce: int = 0, expiry: int = 0,
    chain_id: int = CHAIN_ID, vault: str = VAULT_ADDRESS,
) -> bytes:
    return typed_digest(
        vault_domain(chain_id, vault),
        struct_hash(
            RESOLUTION_TYPE,
            word_bytes32(deal_id32), word_bytes32(terms_hash_ref),
            word_bytes32(decision_id32(decision_id)), word_int(seller_bps),
            word_bytes32(receipt_hash), word_int(nonce), word_int(expiry),
        ),
    )

File: examples/data-seller-agent/src/settlement.ts
/**
 * §4.4 settlement profile (`fixed_outcome/v1`): the EIP-712 constructions the
 * Coordination Engine and the EscrowVault verify.
 *
 * All eight struct type strings are implemented and verified against the full
 * `vectors/v1/settlement` set — the seller *signs* only its own subset
 * (Agreement, Activation co-sign, Delivery, Appeal, RefundConsent), but an
 * implementation that cannot reproduce every digest has diverged from the
 * vault, whatever else it passes.
 *
 * Mirrors the Python examples' `settlement.py`. Implemented from the spec
 * alone; `@noble` supplies the primitives.
 */
import { keccak_256 } from "@noble/hashes/sha3";
import { concatBytes, fromHex, keccak256, recover, sign, utf8 } from "./signing.js";

/** The two §4.4 domains. `KiteFulfill` (accept gate) has NO verifyingContract. */
export interface Eip712Domain {
  name: string;
  version: string;
  chainId: number;
  verifyingContract?: string;
}

/** Struct type strings — verbatim, frozen (§4.4). */
export const TYPE_STRINGS = {
  Agreement:
    "Agreement(bytes32 agreementId,bytes32 termsHash,uint256 amount,address buyerAgent,address sellerAgent)",
  Activation:
    "Activation(bytes32 termsHash,address buyer,address buyerAgent,address sellerAgent,address sellerPayout,address arbiter,uint256 amount,uint64 fundingDeadline,uint64 deliveryWindow,uint64 deliveryConfirmationWindow,uint64 appealResponseWindow,uint64 arbitrationWindow)",
  Delivery:
    "Delivery(bytes32 dealId,bytes32 termsHash,bytes32 deliveryHash,bytes32 receiptHash,uint64 nonce,uint64 expiry)",
  Acceptance: "Acceptance(bytes32 dealId,bytes32 termsHash,bytes32 receiptHash,uint64 nonce,uint64 expiry)",
  Rejection:
    "Rejection(bytes32 dealId,bytes32 termsHash,bytes32 reasonHash,bytes32 receiptHash,uint64 nonce,uint64 expiry)",
  Appeal: "Appeal(bytes32 dealId,bytes32 termsHash,bytes32 receiptHash,uint64 nonce,uint64 expiry)",
  RefundConsent:
    "RefundConsent(bytes32 dealId,bytes32 termsHash,bytes32 receiptHash,uint64 nonce,uint64 expiry)",
  Resolution:
    "Resolution(bytes32 dealId,bytes32 termsHash,bytes32 decisionId,uint16 sellerBps,bytes32 receiptHash,uint64 nonce,uint64 expiry)",
} as const;

export type SettlementStructName = keyof typeof TYPE_STRINGS;

// ── ABI words ────────────────────────────────────────────────────────────────

function wordUint(value: number | bigint | string): Uint8Array {
  const n = BigInt(value);
  if (n < 0n) throw new Error(`negative uint: ${value}`);
  const out = new Uint8Array(32);
  let x = n;
  for (let i = 31; i >= 0 && x > 0n; i--) {
    out[i] = Number(x & 0xffn);
    x >>= 8n;
  }
  if (x > 0n) throw new Error(`uint overflows 32 bytes: ${value}`);
  return out;
}

/** A bytes32 from either spelling the wire uses: `sha256:<hex>` or `0x<hex>` — 32 raw digest bytes either way. */
function wordBytes32(value: string): Uint8Array {
  const h = value.startsWith("sha256:") ? value.slice(7) : value.startsWith("0x") ? value.slice(2) : value;
  const raw = fromHex(h);
  if (raw.length !== 32) throw new Error(`not 32 bytes: ${value}`);
  return raw;
}

function wordAddress(value: string): Uint8Array {
  const raw = fromHex(value);
  if (raw.length !== 20) throw new Error(`not an address: ${value}`);
  return concatBytes(new Uint8Array(12), raw);
}

// ── §4.4 field derivations — normative ───────────────────────────────────────

/** `agreementId` (bytes32) = keccak256 of the UTF-8 bytes of the Runtime-assigned agreement id string. */
export const agreementIdWord = (agreementId: string): Uint8Array => keccak_256(utf8(agreementId));

/** Vault `reasonHash` = keccak256 of the UTF-8 bytes of the wire's `reasonCode` (§4.2). */
export const reasonHashWord = (reasonCode: string): Uint8Array => keccak_256(utf8(reasonCode));

/** Vault `decisionId` (bytes32) = keccak256 of the UTF-8 bytes of the wire's `decisionId` string (§4.2). */
export const decisionIdWord = (decisionId: string): Uint8Array => keccak_256(utf8(decisionId));

/**
 * The contract's decimal `price.amount` in USDC base units (6 decimals):
 * `"25.00"` → `25000000n`. The conversion happens ONCE, by the party signing
 * the Agreement — the Activation read back through `funding` already carries
 * base units, and converting those again produces a digest 10^6 too large.
 */
export function usdcBaseUnits(decimalAmount: string): bigint {
  const m = /^([0-9]+)(?:\.([0-9]{1,6}))?$/.exec(decimalAmount);
  if (!m) throw new Error(`not a USDC decimal amount: ${decimalAmount}`);
  const whole = m[1]!;
  const frac = (m[2] ?? "").padEnd(6, "0");
  return BigInt(whole) * 1_000_000n + BigInt(frac);
}

// ── EIP-712 ──────────────────────────────────────────────────────────────────

export function domainSeparator(domain: Eip712Domain): Uint8Array {
  const withContract = domain.verifyingContract !== undefined;
  const typeString = withContract
    ? "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
    : "EIP712Domain(string name,string version,uint256 chainId)";
  const words = [
    keccak_256(utf8(domain.name)),
    keccak_256(utf8(domain.version)),
    wordUint(domain.chainId),
  ];
  if (withContract) words.push(wordAddress(domain.verifyingContract!));
  return keccak_256(concatBytes(keccak_256(utf8(typeString)), ...words));
}

interface FieldSpec {
  type: string;
  name: string;
}

function parseTypeString(typeString: string): FieldSpec[] {
  const inner = typeString.slice(typeString.indexOf("(") + 1, typeString.lastIndexOf(")"));
  return inner.split(",").map((f) => {
    const [type, name] = f.trim().split(" ");
    if (!type || !name) throw new Error(`malformed field in type string: "${f}"`);
    return { type, name };
  });
}

/**
 * One ABI word per field, INCLUDING the normative §4.4 derivations. The
 * struct carries the wire's source strings — `agreementId` as the id string,
 * `amount` as either the contract decimal (Agreement, optionally pinned by
 * `amountBaseUnits`) or the read-back base-units integer (Activation),
 * Rejection's `reasonCode`, Resolution's `decisionId` — so a signer that
 * derives differently produces a different digest.
 */
function fieldWord(field: FieldSpec, struct: Record<string, unknown>, typeString: string): Uint8Array {
  if (field.name === "agreementId") return agreementIdWord(String(struct["agreementId"]));
  if (field.name === "reasonHash") return reasonHashWord(String(struct["reasonCode"]));
  if (field.name === "decisionId") return decisionIdWord(String(struct["decisionId"]));
  if (field.name === "amount") {
    // §4.4: where the base-unit conversion happens differs BY STRUCT, never
    // by spelling. The Activation's amount is ALREADY base units (the wire's
    // read-back integer, optionally pinned as `amountBaseUnits`); the
    // Agreement's is the contract's decimal `price.amount`, which the schema
    // also permits in integer form ("25" ≡ "25.00") — a dot-based branch
    // here encoded "25" as 25 base units instead of 25,000,000.
    if (typeString.startsWith("Activation(")) {
      return wordUint(String(struct["amountBaseUnits"] ?? struct["amount"]));
    }
    if (struct["amountBaseUnits"] !== undefined) return wordUint(String(struct["amountBaseUnits"]));
    return wordUint(usdcBaseUnits(String(struct["amount"])));
  }
  const value = struct[field.name];
  if (value === undefined) throw new Error(`struct is missing "${field.name}"`);
  if (field.type === "address") return wordAddress(String(value));
  if (field.type === "bytes32") return wordBytes32(String(value));
  if (field.type.startsWith("uint")) return wordUint(value as number | string);
  throw new Error(`unsupported field type: ${field.type}`);
}

/** keccak256( typeHash ‖ one word per field ). */
export function structHash(typeString: string, struct: Record<string, unknown>): Uint8Array {
  const words = parseTypeString(typeString).map((f) => fieldWord(f, struct, typeString));
  return keccak_256(concatBytes(keccak_256(utf8(typeString)), ...words));
}

/** The EIP-712 signing preimage: 0x1901 ‖ domainSeparator ‖ structHash. keccak256 of this is the digest. */
export function signingPreimage(
  domain: Eip712Domain,
  typeString: string,
  struct: Record<string, unknown>,
): Uint8Array {
  return concatBytes(Uint8Array.of(0x19, 0x01), domainSeparator(domain), structHash(typeString, struct));
}

export function digest(
  domain: Eip712Domain,
  typeString: string,
  struct: Record<string, unknown>,
): Uint8Array {
  return keccak256(signingPreimage(domain, typeString, struct));
}

/** Sign a §4.4 struct with the one runtime key. Same 65-byte spelling as every other signature here. */
export function signStruct(
  privateKey: Uint8Array,
  domain: Eip712Domain,
  typeString: string,
  struct: Record<string, unknown>,
): string {
  return sign(privateKey, signingPreimage(domain, typeString, struct));
}

/**
 * Byte equality for 32-byte anchors: the system mints both spellings
 * (`sha256:<hex>` on the coordination layer, `0x<hex>` on chain-adjacent
 * views), and they name the same value. A literal string comparison here
 * refused every real deal — the Python seller learned this live.
 */
export function bytes32Equal(a: string, b: string): boolean {
  const strip = (s: string) => (s.startsWith("sha256:") ? s.slice(7) : s.startsWith("0x") ? s.slice(2) : s).toLowerCase();
  const ha = strip(a);
  const hb = strip(b);
  return ha.length === 64 && ha === hb;
}

/** Recover the signer of a §4.4 signature, or null if malformed. */
export function recoverStruct(
  signature: string,
  domain: Eip712Domain,
  typeString: string,
  struct: Record<string, unknown>,
): string | null {
  return recover(signature, signingPreimage(domain, typeString, struct));
}
